File: implementations/python/lsm_trees.py
"""
Log-Structured Merge-Tree (LSM-Tree) Implementation in Python 3.

Provides:
- BloomFilter: Probabilistic filter using double-hashing (FNV-1a based).
- SSTable: Sorted String Table with sparse block index and Bloom filter.
- LSMTree: Complete implementation with MemTable, Write-Ahead Log (WAL),
  deletion tombstones, and RocksDB-style Leveled Compaction.
"""

import bisect
import math
import random
import unittest
from typing import Optional, List, Dict, NamedTuple


class Entry(NamedTuple):
    key: str
    value: str
    is_tombstone: bool = False


class BloomFilter:
    """
    Bloom Filter supporting fast negative lookup checks.
    """
    def __init__(self, expected_keys: int, fpp: float = 0.01):
        if expected_keys <= 0:
            expected_keys = 1
        m = -expected_keys * math.log(fpp) / (math.log(2) ** 2)
        self.num_bits = max(64, int(math.ceil(m)))
        self.num_hashes = max(1, int(round((self.num_bits / expected_keys) * math.log(2))))
        self.bits = [0] * ((self.num_bits + 63) // 64)

    def _hash(self, key: str, seed: int) -> int:
        h = 14695981039346656037 ^ seed
        for c in key.encode('utf-8'):
            h ^= c
            h = (h * 1099511628211) & 0xFFFFFFFFFFFFFFFF
        return h

    def add(self, key: str):
        h1 = self._hash(key, 0)
        h2 = self._hash(key, 0x9e3779b97f4a7c15)
        for i in range(self.num_hashes):
            idx = (h1 + i * h2) % self.num_bits
            self.bits[idx // 64] |= (1 << (idx % 64))

    def may_contain(self, key: str) -> bool:
        h1 = self._hash(key, 0)
        h2 = self._hash(key, 0x9e3779b97f4a7c15)
        for i in range(self.num_hashes):
            idx = (h1 + i * h2) % self.num_bits
            if not ((self.bits[idx // 64] >> (idx % 64)) & 1):
                return False
        return True


class SSTable:
    """
    Immutable Sorted String Table component with sparse block indexing.
    """
    INDEX_INTERVAL = 4

    def __init__(self, entries: List[Entry]):
        assert len(entries) > 0
        self.entries = sorted(entries, key=lambda e: e.key)
        self.min_key = self.entries[0].key
        self.max_key = self.entries[-1].key

        self.bloom = BloomFilter(len(self.entries), 0.01)
        self.sparse_index: List[tuple] = []  # (key, offset)

        for i, e in enumerate(self.entries):
            self.bloom.add(e.key)
            if i % self.INDEX_INTERVAL == 0:
                self.sparse_index.append((e.key, i))

    def overlaps_with(self, low: str, high: str) -> bool:
        return not (self.max_key < low or self.min_key > high)

    def get(self, key: str) -> Optional[Entry]:
        if key < self.min_key or key > self.max_key:
            return None
        if not self.bloom.may_contain(key):
            return None

        keys = [item[0] for item in self.sparse_index]
        idx = bisect.bisect_right(keys, key) - 1
        if idx < 0:
            idx = 0

        start = self.sparse_index[idx][1]
        end = self.sparse_index[idx + 1][1] if idx + 1 < len(self.sparse_index) else len(self.entries)

        for i in range(start, end):
            if self.entries[i].key == key:
                return self.entries[i]
            if self.entries[i].key > key:
                break
        return None


class LSMTree:
    """
    Log-Structured Merge-Tree with Leveled Compaction.
    """
    MEMTABLE_THRESHOLD = 8
    L0_COMPACTION_THRESHOLD = 4
    MAX_LEVELS = 4

    def __init__(self):
        self.memtable: Dict[str, Entry] = {}
        self.levels: List[List[SSTable]] = [[] for _ in range(self.MAX_LEVELS)]
        self.wal: List[str] = []

    def put(self, key: str, value: str):
        self.wal.append(f"PUT:{key}={value}")
        self.memtable[key] = Entry(key, value, False)
        if len(self.memtable) >= self.MEMTABLE_THRESHOLD:
            self.flush()

    def remove(self, key: str):
        self.wal.append(f"DEL:{key}")
        self.memtable[key] = Entry(key, "", True)
        if len(self.memtable) >= self.MEMTABLE_THRESHOLD:
            self.flush()

    def flush(self):
        if not self.memtable:
            return

        entries = [self.memtable[k] for k in sorted(self.memtable.keys())]
        sstable = SSTable(entries)
        self.levels[0].insert(0, sstable)
        self.memtable.clear()
        self.wal.clear()

        if len(self.levels[0]) >= self.L0_COMPACTION_THRESHOLD:
            self._compact_level(0)

    def _compact_level(self, level: int):
        if level >= self.MAX_LEVELS - 1:
            return

        if level == 0:
            tables_to_compact = list(self.levels[0])
            low_key = min(t.min_key for t in tables_to_compact)
            high_key = max(t.max_key for t in tables_to_compact)
        else:
            if not self.levels[level]:
                return
            pick = self.levels[level][0]
            tables_to_compact = [pick]
            low_key = pick.min_key
            high_key = pick.max_key

        next_level = level + 1
        next_overlapping = [t for t in self.levels[next_level] if t.overlaps_with(low_key, high_key)]
        next_retained = [t for t in self.levels[next_level] if not t.overlaps_with(low_key, high_key)]

        merged: Dict[str, Entry] = {}
        for t in next_overlapping:
            for e in t.entries:
                merged[e.key] = e

        for t in reversed(tables_to_compact):
            for e in t.entries:
                merged[e.key] = e

        is_bottom = (next_level == self.MAX_LEVELS - 1)
        final_entries: List[Entry] = []
        for k in sorted(merged.keys()):
            e = merged[k]
            if is_bottom and e.is_tombstone:
                continue
            final_entries.append(e)

        if level == 0:
            self.levels[0].clear()
        else:
            self.levels[level].pop(0)

        new_next_tables: List[SSTable] = []
        if final_entries:
            chunk_size = self.MEMTABLE_THRESHOLD * 2
            for i in range(0, len(final_entries), chunk_size):
                chunk = final_entries[i:i + chunk_size]
                new_next_tables.append(SSTable(chunk))

        next_retained.extend(new_next_tables)
        next_retained.sort(key=lambda t: t.min_key)
        self.levels[next_level] = next_retained

        capacity = self.L0_COMPACTION_THRESHOLD * (2 ** next_level)
        if len(self.levels[next_level]) > capacity and next_level < self.MAX_LEVELS - 1:
            self._compact_level(next_level)

    def get(self, key: str) -> Optional[str]:
        # 1. MemTable
        if key in self.memtable:
            e = self.memtable[key]
            return None if e.is_tombstone else e.value

        # 2. Level 0
        for sstable in self.levels[0]:
            res = sstable.get(key)
            if res is not None:
                return None if res.is_tombstone else res.value

        # 3. Levels 1+
        for lvl in range(1, self.MAX_LEVELS):
            if not self.levels[lvl]:
                continue
            low, high = 0, len(self.levels[lvl]) - 1
            while low <= high:
                mid = (low + high) // 2
                table = self.levels[lvl][mid]
                if table.min_key <= key <= table.max_key:
                    res = table.get(key)
                    if res is not None:
                        return None if res.is_tombstone else res.value
                    break
                elif key < table.min_key:
                    high = mid - 1
                else:
                    low = mid + 1

        return None


class TestLSMTree(unittest.TestCase):
    def test_random_differential(self):
        lsm = LSMTree()
        oracle = {}
        rng = random.Random(42)
        keys = [f"key_{i}" for i in range(100)]

        for step in range(2500):
            op = rng.randint(0, 2)
            k = rng.choice(keys)

            if op == 0:
                v = f"val_{step}"
                lsm.put(k, v)
                oracle[k] = v
            elif op == 1:
                lsm.remove(k)
                oracle.pop(k, None)
            else:
                expected = oracle.get(k)
                actual = lsm.get(k)
                self.assertEqual(actual, expected)

        lsm.flush()
        for k in keys:
            self.assertEqual(lsm.get(k), oracle.get(k))


if __name__ == '__main__':
    unittest.main()

File: implementations/cpp/lsm_trees.cpp
/**
 * @file lsm_trees.cpp
 * @brief High-performance C++17 reference implementation of a Log-Structured Merge-Tree (LSM-Tree).
 *
 * Implements modern Leveled Compaction (RocksDB / LevelDB architecture):
 * 1. MemTable: In-memory ordered buffer (std::map) for fast sequential and random writes.
 * 2. Write-Ahead Log (WAL): Simulated append-only log providing crash durability.
 * 3. Immutable SSTables (Sorted String Tables):
 *    - Sorted key-value pairs.
 *    - Sparse block index for fast binary search without loading entire file.
 *    - Bloom filter with double-hashing for O(1) negative lookup pruning.
 *    - Explicit tombstones for deletion semantics.
 * 4. Leveled Compaction Engine:
 *    - Level 0: Holds flushed SSTables with potentially overlapping key ranges.
 *    - Levels 1+: Strictly non-overlapping key ranges per level.
 *    - K-way merge compaction deduplicating keys and purging expired tombstones at bottom level.
 * 5. Lookup Priority Order:
 *    MemTable -> L0 (newest to oldest) -> L1+ (binary search key range routing).
 *
 * Fully compliant with -std=c++17 -O3 -Wall -Wextra -Werror.
 */

#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <optional>
#include <cstdint>
#include <cassert>
#include <cmath>
#include <algorithm>
#include <memory>
#include <random>

namespace dsa {

/**
 * @brief Bloom Filter for fast negative lookups.
 */
class BloomFilter {
private:
    std::vector<uint64_t> bits_;
    size_t num_bits_{0};
    size_t num_hashes_{0};

    static uint64_t fnv1a(const std::string& key, uint64_t seed) noexcept {
        uint64_t hash = 14695981039346656037ULL ^ seed;
        for (char c : key) {
            hash ^= static_cast<uint8_t>(c);
            hash *= 1099511628211ULL;
        }
        return hash;
    }

public:
    BloomFilter() = default;

    BloomFilter(size_t expected_keys, double fpp = 0.01) {
        if (expected_keys == 0) expected_keys = 1;
        double m = -static_cast<double>(expected_keys) * std::log(fpp) / (std::log(2.0) * std::log(2.0));
        num_bits_ = std::max(size_t(64), static_cast<size_t>(std::ceil(m)));
        num_hashes_ = std::max(size_t(1), static_cast<size_t>(std::round((num_bits_ / static_cast<double>(expected_keys)) * std::log(2.0))));

        bits_.assign((num_bits_ + 63) / 64, 0);
    }

    void add(const std::string& key) {
        if (num_bits_ == 0) return;
        uint64_t h1 = fnv1a(key, 0);
        uint64_t h2 = fnv1a(key, 0x9e3779b97f4a7c15ULL);
        for (size_t i = 0; i < num_hashes_; ++i) {
            uint64_t combined = h1 + i * h2;
            size_t bit_idx = combined % num_bits_;
            bits_[bit_idx / 64] |= (1ULL << (bit_idx % 64));
        }
    }

    [[nodiscard]] bool may_contain(const std::string& key) const noexcept {
        if (num_bits_ == 0) return true;
        uint64_t h1 = fnv1a(key, 0);
        uint64_t h2 = fnv1a(key, 0x9e3779b97f4a7c15ULL);
        for (size_t i = 0; i < num_hashes_; ++i) {
            uint64_t combined = h1 + i * h2;
            size_t bit_idx = combined % num_bits_;
            if (!((bits_[bit_idx / 64] >> (bit_idx % 64)) & 1ULL)) {
                return false;
            }
        }
        return true;
    }
};

/**
 * @brief Key-Value entry supporting deletion tombstones.
 */
struct Entry {
    std::string key;
    std::string value;
    bool is_tombstone{false};

    bool operator<(const Entry& other) const noexcept {
        return key < other.key;
    }
};

/**
 * @brief Immutable SSTable (Sorted String Table) component.
 */
class SSTable {
public:
    struct IndexEntry {
        std::string key;
        size_t offset;
    };

private:
    std::vector<Entry> entries_;
    std::vector<IndexEntry> sparse_index_;
    BloomFilter bloom_;
    std::string min_key_;
    std::string max_key_;
    static constexpr size_t INDEX_INTERVAL = 4; // Sample every 4 entries

public:
    explicit SSTable(std::vector<Entry> entries) : entries_(std::move(entries)) {
        assert(!entries_.empty());
        std::sort(entries_.begin(), entries_.end());

        min_key_ = entries_.front().key;
        max_key_ = entries_.back().key;

        bloom_ = BloomFilter(entries_.size(), 0.01);
        for (size_t i = 0; i < entries_.size(); ++i) {
            bloom_.add(entries_[i].key);
            if (i % INDEX_INTERVAL == 0) {
                sparse_index_.push_back({entries_[i].key, i});
            }
        }
    }

    [[nodiscard]] const std::string& min_key() const noexcept { return min_key_; }
    [[nodiscard]] const std::string& max_key() const noexcept { return max_key_; }
    [[nodiscard]] const std::vector<Entry>& entries() const noexcept { return entries_; }
    [[nodiscard]] size_t size() const noexcept { return entries_.size(); }

    [[nodiscard]] bool overlaps_with(const std::string& low, const std::string& high) const noexcept {
        return !(max_key_ < low || min_key_ > high);
    }

    [[nodiscard]] std::optional<Entry> get(const std::string& key) const {
        if (key < min_key_ || key > max_key_) return std::nullopt;
        if (!bloom_.may_contain(key)) return std::nullopt;

        if (sparse_index_.empty()) return std::nullopt;

        size_t low = 0, high = sparse_index_.size() - 1, block_idx = 0;
        while (low <= high) {
            size_t mid = low + (high - low) / 2;
            if (sparse_index_[mid].key <= key) {
                block_idx = mid;
                low = mid + 1;
            } else {
                if (mid == 0) break;
                high = mid - 1;
            }
        }

        size_t start_offset = sparse_index_[block_idx].offset;
        size_t end_offset = (block_idx + 1 < sparse_index_.size()) ? sparse_index_[block_idx + 1].offset : entries_.size();

        for (size_t i = start_offset; i < end_offset; ++i) {
            if (entries_[i].key == key) {
                return entries_[i];
            }
            if (entries_[i].key > key) break;
        }

        return std::nullopt;
    }
};

/**
 * @brief LSM-Tree implementing Leveled Compaction.
 */
class LSMTree {
public:
    static constexpr size_t MEMTABLE_THRESHOLD = 8;
    static constexpr size_t L0_COMPACTION_THRESHOLD = 4;
    static constexpr size_t MAX_LEVELS = 4;

private:
    std::map<std::string, Entry> memtable_;
    std::vector<std::shared_ptr<SSTable>> levels_[MAX_LEVELS];
    std::vector<std::string> wal_log_;

    void flush_memtable() {
        if (memtable_.empty()) return;

        std::vector<Entry> entries;
        entries.reserve(memtable_.size());
        for (auto& kv : memtable_) {
            entries.push_back(kv.second);
        }

        auto sstable = std::make_shared<SSTable>(std::move(entries));
        levels_[0].insert(levels_[0].begin(), sstable);
        memtable_.clear();
        wal_log_.clear();

        if (levels_[0].size() >= L0_COMPACTION_THRESHOLD) {
            compact_level(0);
        }
    }

    void compact_level(size_t level) {
        if (level >= MAX_LEVELS - 1) return;

        std::vector<std::shared_ptr<SSTable>> tables_to_compact;
        std::string low_key, high_key;

        if (level == 0) {
            tables_to_compact = levels_[0];
            low_key = tables_to_compact.front()->min_key();
            high_key = tables_to_compact.front()->max_key();
            for (auto& t : tables_to_compact) {
                low_key = std::min(low_key, t->min_key());
                high_key = std::max(high_key, t->max_key());
            }
        } else {
            if (levels_[level].empty()) return;
            auto pick = levels_[level].front();
            tables_to_compact.push_back(pick);
            low_key = pick->min_key();
            high_key = pick->max_key();
        }

        size_t next_level = level + 1;
        std::vector<std::shared_ptr<SSTable>> next_overlapping;
        std::vector<std::shared_ptr<SSTable>> next_retained;

        for (auto& t : levels_[next_level]) {
            if (t->overlaps_with(low_key, high_key)) {
                next_overlapping.push_back(t);
            } else {
                next_retained.push_back(t);
            }
        }

        std::map<std::string, Entry> merged_entries;

        // Older level first
        for (auto& t : next_overlapping) {
            for (const auto& e : t->entries()) {
                merged_entries[e.key] = e;
            }
        }

        // Newer level overrides
        for (auto it = tables_to_compact.rbegin(); it != tables_to_compact.rend(); ++it) {
            for (const auto& e : (*it)->entries()) {
                merged_entries[e.key] = e;
            }
        }

        bool is_bottom_level = (next_level == MAX_LEVELS - 1);
        std::vector<Entry> final_entries;
        final_entries.reserve(merged_entries.size());
        for (auto& kv : merged_entries) {
            if (is_bottom_level && kv.second.is_tombstone) {
                continue; // Discard expired tombstone at bottom level
            }
            final_entries.push_back(kv.second);
        }

        if (level == 0) {
            levels_[0].clear();
        } else {
            levels_[level].erase(levels_[level].begin());
        }

        std::vector<std::shared_ptr<SSTable>> new_next_tables;
        if (!final_entries.empty()) {
            size_t chunk_size = MEMTABLE_THRESHOLD * 2;
            for (size_t i = 0; i < final_entries.size(); i += chunk_size) {
                size_t end = std::min(i + chunk_size, final_entries.size());
                std::vector<Entry> chunk(final_entries.begin() + i, final_entries.begin() + end);
                new_next_tables.push_back(std::make_shared<SSTable>(std::move(chunk)));
            }
        }

        next_retained.insert(next_retained.end(), new_next_tables.begin(), new_next_tables.end());
        std::sort(next_retained.begin(), next_retained.end(), [](const auto& a, const auto& b) {
            return a->min_key() < b->min_key();
        });

        levels_[next_level] = std::move(next_retained);

        size_t capacity = L0_COMPACTION_THRESHOLD * (1 << next_level);
        if (levels_[next_level].size() > capacity && next_level < MAX_LEVELS - 1) {
            compact_level(next_level);
        }
    }

public:
    LSMTree() = default;

    void put(const std::string& key, const std::string& value) {
        wal_log_.push_back("PUT:" + key + "=" + value);
        memtable_[key] = {key, value, false};
        if (memtable_.size() >= MEMTABLE_THRESHOLD) {
            flush_memtable();
        }
    }

    void remove(const std::string& key) {
        wal_log_.push_back("DEL:" + key);
        memtable_[key] = {key, "", true};
        if (memtable_.size() >= MEMTABLE_THRESHOLD) {
            flush_memtable();
        }
    }

    [[nodiscard]] std::optional<std::string> get(const std::string& key) {
        // Priority 1: MemTable
        auto it = memtable_.find(key);
        if (it != memtable_.end()) {
            if (it->second.is_tombstone) return std::nullopt;
            return it->second.value;
        }

        // Priority 2: Level 0 (ordered newest to oldest)
        for (auto& sstable : levels_[0]) {
            auto res = sstable->get(key);
            if (res.has_value()) {
                if (res->is_tombstone) return std::nullopt;
                return res->value;
            }
        }

        // Priority 3: Levels 1+ (disjoint key ranges per level)
        for (size_t lvl = 1; lvl < MAX_LEVELS; ++lvl) {
            if (levels_[lvl].empty()) continue;

            size_t low = 0, high = levels_[lvl].size() - 1;
            while (low <= high) {
                size_t mid = low + (high - low) / 2;
                const auto& table = levels_[lvl][mid];
                if (key >= table->min_key() && key <= table->max_key()) {
                    auto res = table->get(key);
                    if (res.has_value()) {
                        if (res->is_tombstone) return std::nullopt;
                        return res->value;
                    }
                    break;
                } else if (key < table->min_key()) {
                    if (mid == 0) break;
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            }
        }

        return std::nullopt;
    }

    void flush() {
        flush_memtable();
    }
};

} // namespace dsa

int main() {
    using namespace dsa;
    std::cout << "Testing LSM-Tree with Leveled Compaction against std::map oracle...\n";

    LSMTree lsm;
    std::map<std::string, std::string> oracle;

    std::mt19937 rng(1337);
    std::vector<std::string> keys;
    for (int i = 0; i < 250; ++i) {
        keys.push_back("usr_" + std::to_string(i));
    }

    for (int step = 0; step < 5000; ++step) {
        int op = rng() % 3;
        const std::string& k = keys[rng() % keys.size()];

        if (op == 0) { // Put
            std::string v = "payload_" + std::to_string(step);
            lsm.put(k, v);
            oracle[k] = v;
        } else if (op == 1) { // Remove
            lsm.remove(k);
            oracle.erase(k);
        } else { // Get
            auto expected = oracle.count(k) ? std::make_optional(oracle[k]) : std::nullopt;
            auto actual = lsm.get(k);
            assert(actual == expected);
        }
    }

    lsm.flush();
    for (const auto& k : keys) {
        auto expected = oracle.count(k) ? std::make_optional(oracle[k]) : std::nullopt;
        auto actual = lsm.get(k);
        assert(actual == expected);
    }

    std::cout << "All 5,000 LSM-Tree operations with Leveled Compaction passed with 100% precision!\n";
    return 0;
}
